AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/SQLite/Statement+Model.swift

```diff
@@ -35,99 +33,159 @@ protocol StatementModelConvertible {
+                    modelDictionary.updateValue(lazyList, forKeyPath: listKeyPath)
+                }
             }
         }
+        // swiftlint:disable:next force_cast
+        return modelDictionary as! ModelValues
+    }
 
-        if let id = values[schema.primaryKey.name] as? String {
-            // create instances of lazy-load lists of fields that represent "many" associations
-            schema.fields.values.filter { $0.isArray }.forEach { field in
-                // TODO extract this to a List utility
-                let lazyListDecodable: [String: Any?] = [
-                    "associatedId": id,
-                    "associatedField": field.associatedField?.name,
-                    "elements": []
-                ]
-                values[field.name] = lazyListDecodable
+}
+
+internal extension List {
+
+    /// Creates a data structure that is capable of initializing a `List<M>` with
+    /// lazy-load capabilities when the list is being decoded.
+    ///
+    /// See the `List.init(from:Decoder)` for details.
+    static func lazyInit(associatedId: String, associatedWith: String?) -> [String: Any?] {
+        return [
+            "associatedId": associatedId,
+            "associatedField": associatedWith,
+            "elements": []
+        ]
+    }
+}
+
+internal extension Dictionary where Key == String, Value == Any? {
+
+    /// Utility to create a `NSMutableDictionary` from a Swift `Dictionary<String, Any?>`.
+    func mutableCopy() -> NSMutableDictionary {
+        // swiftlint:disable:next force_cast
+        return (self as NSDictionary).mutableCopy() as! NSMutableDictionary
+    }
+}
+
+/// Extension that adds utilities to created nested values in a dictionary
+/// from a `keyPath` notation (e.g. `root.with.some.nested.prop`.
+internal extension NSMutableDictionary {
+
+    /// Utility to allows Swift standard types to be used in `setObject`
+    /// of the `NSMutableDictionary`.
+    ///
+    /// - Parameters:
+    ///   - value: the value to be set
+    ///   - key: the key as a `String`
+    func updateValue(_ value: Value?, forKey key: String) {
+        setObject(value as Any, forKey: key as NSString)
+    }
+
+    /// Utility that enables the automatic creation of nested dictionaries when
+    /// a `keyPath` is passed, even if no existing value is set in that `keyPath`.
+    ///
+    /// This function will auto-create nested structures and set the value accordingly.
+    ///
+    /// - Example
+    ///
+    /// ```swift
+    /// let dict = [:].mutableCopy()
+    /// dict.updateValue(1, "some.nested.value")
+    ///
+    /// // dict now is
+    /// [
+    ///     "some": [
+    ///         "nested": [
+    ///             "value": 1
+    ///         ]
+    ///     ]
+    /// ]
+    /// ```
+    ///
+    /// - Parameters:
+    ///   - value: the value to be set
+    ///   - keyPath: the key path as a `String` (e.g. "nested.key")
+    func updateValue(_ value: Value?, forKeyPath keyPath: String) {
```

Comment:
Hmm stupid question: what if we have something like `nested..key`. will `keyComponents` be 2 or 3?  Maybe there's some extra sanity checks on the data that is being passed in here?

Reply:
that's an invalid `keyPath`. We could add sanity checks, but it wouldn't make a difference because this is used in a controlled environment. The `keyPath` passed here is created by `SelectStatement`, is select statement creates an invalid key, those tests will fail. If tests don't catch it, the SQL operation will fail before it gets here.

but good question, I don't know how https://developer.apple.com/documentation/objectivec/nsobject/1418139-setvalue behaves in those cases. It would be good to mimic its behavior